File: imp-haskell/src/Ast.hs
{-# LANGUAGE GADTs #-}
module Ast where

type Name = String

data AExp where
  NUM :: Int -> AExp
  LOC :: Name -> AExp
  ADD :: AExp -> AExp -> AExp
  SUB :: AExp -> AExp -> AExp
  MUL :: AExp -> AExp -> AExp
  deriving (Show)

data BExp where
  TRUE  :: BExp
  FALSE :: BExp
  EQUAL :: BExp -> BExp -> BExp
  LEQ   :: BExp -> BExp -> BExp
  NOT   :: BExp -> BExp
  AND   :: BExp -> BExp -> BExp
  OR    :: BExp -> BExp -> BExp
  deriving (Show)

data Command where
  SKIP   :: Command
  ASSIGN :: Name -> AExp -> Command
  SEQ    :: Command -> Command -> Command
  IF     :: BExp -> Command -> Command -> Command
  WHILE  :: BExp -> Command -> Command
  deriving (Show)

File: imp-ocaml/ast.ml
type var = string

type aexp =
  | Num : int -> aexp
  | Var : var -> aexp
  | Add : aexp * aexp -> aexp
  | Sub : aexp * aexp -> aexp
  | Mul : aexp * aexp -> aexp

type bexp =
  | True  : bexp
  | False : bexp
  | Equal : bexp * bexp -> bexp
  | Leq   : bexp * bexp -> bexp
  | Not   : bexp -> bexp
  | And   : bexp * bexp -> bexp
  | Or    : bexp * bexp -> bexp

type command =
  | Skip   : command
  | Assign : var * aexp -> command
  | Seq    : command * command -> command
  | If     : bexp * command * command -> command
  | While  : bexp * command -> command

